modules/data.atmosphere/R/met_temporal_downscale.Gaussian_ensemble.R

```diff
@@ -0,0 +1,398 @@
+    
+    # Temperature For temperature, we create a random normal distribution with the
+    # mean of the source value and the standard deviation of the training value
+    # within the window, and step through that window sequentially so that we are
+    # only using the standard devaition of values recorded at the same time as the
+    # value we are generating.
+    temp <- c("air_temperature", "air_temperature_max", "air_temperature_min")
+    df <- data.frame()
+    for (u in temp) {
+      woo <- vector()
+      a <- as.numeric(train[[u]])
+      sour <- as.numeric(source[[u]])
+      if (all(is.na(sour)) == TRUE) {
+        woo <- rep(NA, reso_len)
+      }
+      if (all(is.na(sour)) == FALSE) {
+        for (x in seq_along(sour)) {
+          four <- vector()
+          for (n in seq_along(1:div)) {
+            lowday <- x * div - wd * sd_step + n
+            highday <- x * div + wd * sd_step + n
+            if (lowday < 0) {
+              lowday <- x + n
+              highday <- x + n + wd * sd_step + n
+            }
+            if (highday > reso_len) {
+              highday <- reso_len
+              lowday <- x * div - wd * sd_step + n
+            }
+            four[n] <- rnorm(1, mean = sour[x], sd = sd(a[seq(from = lowday, 
+                                                              to = highday, by = sd_step)]))
+          }
+          woo <- append(woo, four)
+        }
+      }
+      df[1:length(woo), u] <- woo
+    }
+    
+    # Makes sure that max temps >= avg temp and min temps <= avg temp
+    for (x in seq_along(df$air_temperature)) {
+      if (all(is.na(df$air_temperature_max)) == FALSE) {
+        if (df$air_temperature_max[x] < df$air_temperature[x]) {
+          df$air_temperature_max[x] <- df$air_temperature[x]
+        }
+      }
+      if (all(is.na(df$air_temperature_max)) == TRUE) {
+        df$air_temperature_max <- rep(NA, reso_len)
+      }
+      if (all(is.na(df$air_temperature_min)) == FALSE) {
+        if (df$air_temperature_min[x] > df$air_temperature[x]) {
+          df$air_temperature_min[x] <- df$air_temperature[x]
+        }
+      }
+      if (all(is.na(df$air_temperature_min)) == TRUE) {
+        df$air_temperature_min <- rep(NA, reso_len)
+      }
+    }
+    
+    # Precipitation_flux this takes the daily total of precipitation and uses that as
+    # a total possible amount of precip.  It randomly distributes the values of
+    # precipitation
+    rand_vect_cont <- function(N, M, sd = 1) {
+      vec <- rnorm(N, M/N, sd)
+      vec/sum(vec) * M
+    }
+    precip <- vector()
+    for (x in seq_along(source$precipitation_flux)) {
+      lowday <- (x - wd) * div
+      highday <- (x + wd) * div
+      if (lowday < 0) {
+        lowday <- 0
+      }
+      if (highday > reso_len) {
+        highday <- reso_len
+      }
+      four <- vector()
+      four <- rand_vect_cont(div, source$precipitation_flux[x], sd = sd(train$precipitation_flux[lowday:highday]))
```

Comment:
Very similar issues as the downscaling of T. Furthermore, for precip you definitely don't want to use an untransformed Gaussian as it will result in massive amounts of negative precip (e.g. most days will be mean 0, which means half the simulated precip will be negative). I see you truncate the negative values below, but by doing that truncation without accounting for it in your mean and SD you're introducing a significant bias. For example, for a very dry season the observed mean has to be positive because it's a sum of a lot of zeros and a few small values. But to reproduce a timeseries that's mostly 0's and a few small values you need the (latent) mean of the Gaussian you're sampling from to be negative.